torch/csrc/autograd/python_variable_indexing.cpp

```diff
@@ -0,0 +1,293 @@
+    } else {
+      invalid_index(obj);
+    }
+  }
+  return result;
+}
+
+static std::vector<Tensor> asTensorList(const variable_list& v) {
+  return std::vector<Tensor>(v.begin(), v.end());
+}
+
+static Variable dispatch_index(const Variable& self, const variable_list& indices) {
+  AutoNoGIL no_gil;
+  AutoGPU auto_gpu(self);
+  return self.index(asTensorList(indices));
+}
+
+static Variable dispatch_index_put_(Variable& self, const variable_list& indices, const Variable& value) {
+  AutoNoGIL no_gil;
+  AutoGPU auto_gpu(self);
+  return self.index_put_(asTensorList(indices), value);
+}
+
+static bool treatSequenceAsTuple(PyObject* index) {
+  if (PyTuple_Check(index)) {
+    return true;
+  }
+  if (!PySequence_Check(index)) {
+    return false;
+  }
+  // This uses a heuristics from NumPy for determining whether to treat
+  // non-tuple sequences as if they were a tuple. From the NumPy code comments:
+  //
+  // "At this point, we're left with a non-tuple, non-array, sequence:
+  //  typically, a list. We use some somewhat-arbitrary heuristics from here
+  //  onwards to decided whether to treat that list as a single index, or a
+  //  list of indices. Backwards compatibility only takes effect for short
+  //  sequences - otherwise we treat it like any other scalar."
+  auto n = PySequence_Size(index);
+  if (n < 0) {
+    PyErr_Clear();
+    return false;
+  }
+  if (n >= 32) {
+    return false;
+  }
+  for (Py_ssize_t i = 0; i < n; i++) {
+    auto obj = THPObjectPtr{PySequence_GetItem(index, i)};
+    if (!obj.get()) {
+      PyErr_Clear();
+      return false;
+    }
+    if (THPVariable_Check(obj.get()) || PySequence_Check(obj.get()) || PySlice_Check(obj.get())) {
+      return true;
+    }
+    if (obj.get() == Py_Ellipsis || obj.get() == Py_None) {
+      return true;
+    }
+  }
+  return false;
+}
+
+static THPObjectPtr wrapTuple(PyObject* index) {
+  THPObjectPtr res;
+  if (treatSequenceAsTuple(index)) {
+    res = PySequence_Tuple(index);
+  } else {
+    res = PyTuple_Pack(1, index);
+  }
+  if (!res) throw python_error();
+  return res;
+}
+
+PyObject* THPVariable_getitem(PyObject* self, PyObject* index) {
+  HANDLE_TH_ERRORS
+  auto& self_ = reinterpret_cast<THPVariable*>(self)->cdata;
+
+  // handle simple types: integers, slices, ellipsis
+  if (index == Py_None) {
+    return wrap(self_.unsqueeze(0));
+  } else if (index == Py_Ellipsis) {
+    return wrap(at::alias(self_));
+  } else if (THPUtils_checkLong(index)) {
+    return wrap(applySelect(self_, 0, THPUtils_unpackLong(index)));
+  } else if (PyBool_Check(index)) {
+    if (index == Py_True) {
+      return wrap(self_.unsqueeze(0));
+    } else {
+      return wrap(self_.type().tensor({0}));
+    }
+  } else if (PySlice_Check(index)) {
+    return wrap(applySlice(self_, 0, index, true));
+  }
+
+  // wrap index in a tuple if it's not already one
+  THPObjectPtr holder = wrapTuple(index);
+
+  variable_list variableIndices;
```

Comment:
I'm making the assumption that this section constitutes advanced indexing, but it would be helpful to have a code comment indicating that this is where those semantics begin.